lightningd/jsonrpc.c

```diff
@@ -575,14 +578,140 @@ struct json_stream *json_stream_fail(struct command *cmd,
+	 * If they completed it, it's freed already. */
+	if (res == &pending)
+		assert(cmd->pending);
+
+	list_for_each(&jcon->commands, cmd, list)
+		assert(cmd->pending);
+
+	return res;
+}
+
+/* A plugin hook to take over (fail/alter) RPC commands */
+struct rpc_command_hook_payload {
+	struct command *cmd;
+	const char *buffer;
+	const jsmntok_t *request;
+};
+
+static void rpc_command_hook_serialize(struct rpc_command_hook_payload *p,
+                                       struct json_stream *s)
+{
+	const jsmntok_t *method, *params;
+
+	method = json_get_member(p->buffer, p->request, "method");
+	params = json_get_member(p->buffer, p->request, "params");
+
+	json_object_start(s, "rpc_command");
+	json_add_string(s, "method", tal_fmt(tmpctx, "%.*s",
+	                                     method->end - method->start,
+	                                     p->buffer + method->start));
+	json_add_string(s, "params", tal_fmt(tmpctx, "%.*s",
+	                                     params->end - params->start,
+	                                     p->buffer + params->start));
+	json_object_end(s);
+}
+
+static void
+rpc_command_hook_callback(struct rpc_command_hook_payload *p,
+                          const char *buffer, const jsmntok_t *resulttok)
+{
+	const jsmntok_t *tok, *method, *params, *custom_return;
+	struct json_stream *response;
+
+	params = json_get_member(p->buffer, p->request, "params");
+
+	/* If no plugin registered, there is no restriction. */
+	if (buffer == NULL || json_tok_streq(buffer, resulttok, "continue"))
+		return was_pending(command_exec(p->cmd->jcon, p->cmd, p->buffer,
+		                                p->request, params));
+
+	/* If the registered plugin did not respond with continue,
+	 * it wants either to replace the request... */
+	tok = json_get_member(buffer, resulttok, "replace");
+	if (tok) {
+		method = json_get_member(buffer, tok, "method");
+		params = json_get_member(buffer, tok, "params");
+		if (!method || !params)
+			return was_pending(command_fail(p->cmd, JSONRPC2_INVALID_REQUEST,
+			                                "Bad response to 'rpc_command' hook: "
+			                                "the 'replace' object must contain a "
+			                                "'method' and a 'params' field."));
+		p->cmd->json_cmd = find_cmd(p->cmd->ld->jsonrpc, buffer, method);
+		return was_pending(command_exec(p->cmd->jcon, p->cmd, buffer,
+		                                method, params));
+	}
+
+	/* ...or return a custom JSONRPC response. */
+	tok = json_get_member(buffer, resulttok, "return");
+	if (tok) {
+		custom_return = json_get_member(buffer, tok, "result");
+		if (custom_return) {
+			response = json_start(p->cmd);
+			json_add_tok(response, "result", custom_return, buffer);
+			json_object_compat_end(response);
+			return was_pending(command_raw_complete(p->cmd, response));
+		}
+
+		custom_return = json_get_member(buffer, tok, "error");
+		if (custom_return) {
```

Comment:
This is a bit strange since it allows us to return both an `error` as well as a `result`. Maybe we shouldn't perform any validation at all and just return the payload of `return` to the caller?

The only issue I see is the need to rewrite the ID back

Reply:
> This is a bit strange since it allows us to return both an error as well as a result

If there is a `result` field in the response we return ? Line 670

> Maybe we shouldn't perform any validation at all and just return the payload of return to the caller?
> The only issue I see is the need to rewrite the ID back

I think that was to be able to replace the call, as https://github.com/ElementsProject/lightning/pull/2925#discussion_r312381909